[REVIEW] Port unary cast to libcudf++

trxcllnt opened the pull request:
Port unaryops `cast` for issue #2950, related to @codereport's PR https://github.com/rapidsai/cudf/pull/3214.

trxcllnt commented:
rerun tests

jrhemstad commented on cpp/include/cudf/unary.hpp:
```diff
@@ -16,8 +16,9 @@
-#include <cudf/cudf.h>
+#include <memory>
 #include <cudf/types.hpp>
+#include <rmm/mr/default_memory_resource.hpp>
```
The forward declarations in `types.hpp` should be sufficient. 
```suggestion
```

jrhemstad commented on cpp/src/unary/cast_ops.cu:
```diff
@@ -0,0 +1,123 @@
+  auto null_mask = copy_bitmask(input, 0, mr);
+  auto output = std::make_unique<column>(
+      type, size, rmm::device_buffer{size * cudf::size_of(type), 0, mr},
+      null_mask, input.null_count(), std::vector<std::unique_ptr<column>>{});
```
Careful, this was deep copying the device buffer.
```suggestion
      std::move(null_mask), input.null_count(), std::vector<std::unique_ptr<column>>{});
```

trxcllnt replied:
Don't we want to deep-copy the device buffer? What happens if the input column is still valid and in use elsewhere?

jrhemstad commented on cpp/src/unary/cast_ops.cu:
```diff
@@ -0,0 +1,123 @@
+  auto null_mask = copy_bitmask(input, 0, mr);
+  auto output = std::make_unique<column>(
+      type, size, rmm::device_buffer{size * cudf::size_of(type), 0, mr},
+      null_mask, input.null_count(), std::vector<std::unique_ptr<column>>{});
```
The default arg should be sufficient. 
```suggestion
      null_mask, input.null_count());
```

jrhemstad commented on cpp/src/unary/cast_ops.cu:
```diff
@@ -0,0 +1,123 @@
+            typename std::enable_if_t<!cudf::is_timestamp<T>() &&
+                                      !cudf::is_numeric<T>()>* = nullptr>
+  void operator()(cudaStream_t stream) {
+    CUDF_FAIL("Column type must be numeric");
```
```suggestion
    CUDF_FAIL("Column type must be numeric or timestamp");
```

trxcllnt commented:
@jrhemstad I cleaned up the implementations and changed the signatures to return `unique_ptr<column>` now that https://github.com/rapidsai/cudf/pull/3232 is merged, I was thinking of updating this PR to call `to_timestamps` and `from_timestamps` when casting TS to/from strings. Does that sound like the right approach?

jrhemstad commented:
> I was thinking of updating this PR to call `to_timestamps` and `from_timestamps` when casting TS to/from strings. Does that sound like the right approach?

If that's the expected behavior of casting to/from strings, then sure.

trxcllnt commented:
It would be convenient to have a single `cudf::cast()` method to convert columns to another type, but after looking through the methods in `cudf/strings/convert`, I've decided against it.

Each string method takes different parameters based on the to/from type, and it isn't possible to define `cast()` overloads for each convert func parameter list. Better to guide people to those functions explicitly when they want to go to/from strings.

harrism commented:
What's the status of this PR?

jrhemstad commented on cpp/include/cudf/unary.hpp:
```diff
@@ -42,5 +42,17 @@ std::unique_ptr<cudf::column> is_null(cudf::column_view const& input);
 
+/**
+ * @brief  Casts data from dtype specified in input to dtype specified in output
+ *
```
Might want to add some documentation about what types are supported.

trxcllnt commented:
@harrism to answer your question, now that I've updated the `cast` docstring I think this is ready to merge.
